### PR TITLE
fix: convert_to_genai_function_declarations

### LINKED ISSUE #234: [bugifx] ShellTool not work for gemini-1.5 Search because of TYPE_ENUM[v["type"]] returns KeyError: 'type'
## Example Code

```python
from dotenv import load_dotenv
from langchain import hub
from langchain.agents import AgentExecutor, create_tool_calling_agent
from langchain_anthropic import ChatAnthropic
from langchain_community.tools.shell.tool import ShellInput, ShellTool
from langchain_core.utils.function_calling import format_tool_to_openai_function
from langchain_google_genai import ChatGoogleGenerativeAI

load_dotenv(verbose=True, override=False)

is_anthropic = False
if is_anthropic:
    model_anthropic = "claude-3-haiku-20240307"
    llm_chat = ChatAnthropic(
        model_name=model_anthropic,
    )
else:
    model_google = "gemini-1.5-pro-latest"
    model_google = "gemini-1.5-flash-latest"
    # model_google = "gemini-1.0-pro" # not accept instruction message.
    llm_chat = ChatGoogleGenerativeAI(
        model=model_google,
    )

shell_tool = ShellTool()
tools = [shell_tool]
llm_chat_with_tools = llm_chat.bind_tools(tools)  # not impl yet

question = "please run pwd command."
agent_scratchpad = []

prompt = hub.pull("hwchase17/openai-functions-agent")
agent = create_tool_calling_agent(llm_chat, tools, prompt)
agent_executor = AgentExecutor(
    agent=agent,
    tools=tools,
    allowed_tools=[shell_tool.name],
)

input_message = {"input": question, "agent_scratchpad": agent_scratchpad}
result = agent_executor.invoke(input_message)

print("result=", result)
```

## Error Message and Stack Trace

```
Traceback (most recent call last):
  File "/home/jinno/git/drill/gamebook/codeinterpreter_api_agent/../test/test_shell_tool.py", line 38, in <module>
    agent = create_tool_calling_agent(llm_chat, tools, prompt)
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain/agents/tool_calling_agent/base.py", line 88, in create_tool_calling_agent
    llm_with_tools = llm.bind_tools(tools)
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/chat_models.py", line 906, in bind_tools
    genai_tools = [tool_to_dict(convert_to_genai_function_declarations(tools))]
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 65, in convert_to_genai_function_declarations
    return convert_to_genai_function_declarations({"function_declarations": tool})
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 68, in convert_to_genai_function_declarations
    function_declarations=[
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 69, in <listcomp>
    _convert_to_genai_function(fc) for fc in tool["function_declarations"]
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 105, in _convert_to_genai_function
    return _convert_tool_to_genai_function(fc)
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 137, in _convert_tool_to_genai_function
    "properties": {
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 139, in <dictcomp>
    "type_": TYPE_ENUM[v["type"]],
KeyError: 'type'
Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/exceptiongroup/_formatting.py", line 71, in exceptiongroup_excepthook
TypeError: 'NoneType' object is not callable
```

## Description

ChatAnthropic works fine.
But ChatGoogleGenerativeAI shows error. It looks be caused by here.

1. TYPE_ENUM[v["type"]] is no care for None case in [here](https://github.com/langchain-ai/langchain-google/blob/2f7f46c3f0e708f55aea30233f10f6821b55e237/libs/genai/langchain_google_genai/_function_utils.py#L139).
2. Return type of [here](https://github.com/langchain-ai/langchain-google/blob/2f7f46c3f0e708f55aea30233f10f6821b55e237/libs/genai/langchain_google_genai/_function_utils.py#L63C16-L63C47) is wrong. 

## System Info

Ubuntu 23.04

langchain==0.2.0
langchain-anthropic==0.1.11
langchain-community==0.0.34
langchain-core==0.2.0
langchain-experimental==0.0.57
langchain-google-genai==1.0.4
langchain-groq==0.1.3
langchain-openai==0.1.6
langchain-text-splitters==0.2.0
langchainhub==0.1.15

### PR DESCRIPTION
**Pull request by nobu007**
Fix #234.

I'm not sure the root cause.
_function_utils.py is refactored with investigating.

Sorry, I didn't care vertexai side because no environment with me.

### before

```
(gamebook) jinno@jinno-desktop:~/git/drill/gamebook/codeinterpreter_api_agent$ python ../test/test_shell_tool.py 
・・・
  File "/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_google_genai/_function_utils.py", line 139, in <dictcomp>
    "type_": TYPE_ENUM[v["type"]],
KeyError: 'type'
```

### after(with debug log)

```
(gamebook) jinno@jinno-desktop:~/git/drill/gamebook/codeinterpreter_api_agent$ python ../test/test_shell_tool.py 
_convert_pydantic_to_genai_function schema= {'title': 'ShellInput', 'description': 'Commands for the Bash Shell tool.', 'type': 'object', 'properties': {'commands': {'title': 'Commands', 'description': 'List of shell commands to run. Deserialized using json.loads', 'anyOf': [{'type': 'string'}, {'type': 'array', 'items': {'type': 'string'}}]}}, 'required': ['commands']}
_get_type_from_schema schema= {'title': 'Commands', 'description': 'List of shell commands to run. Deserialized using json.loads', 'anyOf': [{'type': 'string'}, {'type': 'array', 'items': {'type': 'string'}}]}
_get_type_from_schema schema= {'type': 'string'}
_get_type_from_schema schema= {'type': 'array', 'items': {'type': 'string'}}
types= [<Type.STRING: 1>, <Type.ARRAY: 5>]
_convert_pydantic_to_genai_function function_declaration= name: "terminal"
description: "Commands for the Bash Shell tool."
parameters {
  type_: OBJECT
  properties {
    key: "commands"
    value {
      type_: ARRAY
      description: "List of shell commands to run. Deserialized using json.loads"
    }
  }
  required: "commands"
}

_convert_pydantic_to_genai_function schema= {'title': 'ShellInput', 'description': 'Commands for the Bash Shell tool.', 'type': 'object', 'properties': {'commands': {'title': 'Commands', 'description': 'List of shell commands to run. Deserialized using json.loads', 'anyOf': [{'type': 'string'}, {'type': 'array', 'items': {'type': 'string'}}]}}, 'required': ['commands']}
_get_type_from_schema schema= {'title': 'Commands', 'description': 'List of shell commands to run. Deserialized using json.loads', 'anyOf': [{'type': 'string'}, {'type': 'array', 'items': {'type': 'string'}}]}
_get_type_from_schema schema= {'type': 'string'}
_get_type_from_schema schema= {'type': 'array', 'items': {'type': 'string'}}
types= [<Type.STRING: 1>, <Type.ARRAY: 5>]
_convert_pydantic_to_genai_function function_declaration= name: "terminal"
description: "Commands for the Bash Shell tool."
parameters {
  type_: OBJECT
  properties {
    key: "commands"
    value {
      type_: ARRAY
      description: "List of shell commands to run. Deserialized using json.loads"
    }
  }
  required: "commands"
}

/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_community/tools/shell/tool.py:32: UserWarning: The shell tool has no safeguards by default. Use at your own risk.
  warnings.warn(
Executing command:
 ['pwd']
/home/jinno/.local/share/virtualenvs/gamebook-7W9hF6PY/lib/python3.10/site-packages/langchain_community/tools/shell/tool.py:32: UserWarning: The shell tool has no safeguards by default. Use at your own risk.
  warnings.warn(
Executing command:
 ['pwd']
result= {'input': 'please run pwd command.', 'agent_scratchpad': [], 'output': 'The current working directory is `/home/jinno/git/drill/gamebook/codeinterpreter_api_agent`. \n'}
```

### Note

1. This error also remained in my environment.
https://github.com/google-gemini/generative-ai-python/issues/207

2. libs/genai/langchain_google_genai/_function_utils.py still some problem remains.

- function_declarations inf GoogleTool is not set correctry.
- Maybe _convert_to_genai_function() should return function_declarations not function_declaration.
- But I had no time to care this.

### CODE REVIEW

File: libs/genai/langchain_google_genai/_function_utils.py [hunk 1]
```diff
@@ -160,17 +149,17 @@ def _convert_tool_to_genai_function(tool: BaseTool) -> FunctionDeclaration:
 
 
 def _convert_pydantic_to_genai_function(
-    pydantic_model: Type[BaseModel],
+    pydantic_model: Type[BaseModel], tool_name: str = ""
```
Comment (lkuligin) [inline libs/genai/langchain_google_genai/_function_utils.py:152]: nits: can we declare tool_name as `Optional[str]=None`

File: libs/genai/langchain_google_genai/_function_utils.py [hunk 2]
```diff
@@ -47,29 +42,36 @@ class _ToolDict(TypedDict):
...
-        _ToolDict,
-        _FunctionDeclarationLike,
-        Sequence[_FunctionDeclarationLike],
-    ],
+    tool: GeminiToolLike,
 ) -> GoogleTool:
     if isinstance(tool, GoogleTool):
         return cast(GoogleTool, tool)
-    if isinstance(tool, type) and issubclass(tool, BaseModel):
+    if isinstance(tool, (BaseTool, Type[BaseModel], dict, FunctionDeclaration)):
```
Comment (lkuligin) [inline libs/genai/langchain_google_genai/_function_utils.py:57]: I believe, we want to distinguish a case when dict is OpenAPI spec vs dict is like {"function_declarations": ...}.

**Comment by lkuligin** [comment]
@nobu007 thanks a lot for your contribution! could you check my comments, and also take a look at the linter errors, please.

**Comment by nobu007** [comment]
I just rebased.

File: libs/genai/langchain_google_genai/_function_utils.py [hunk 3]
```diff
@@ -47,29 +42,36 @@ class _ToolDict(TypedDict):
...
-        return _convert_tool_to_genai_function(callable_as_lc_tool()(tool))
-    if isinstance(tool, list):
-        return convert_to_genai_function_declarations({"function_declarations": tool})
+        return GoogleTool(
+            function_declarations=_convert_tool_to_genai_function(
+                callable_as_lc_tool()(tool)
+            )
+        )
+    # can not reach this code?
     if isinstance(tool, dict) and "function_declarations" in tool:
```
Comment (lkuligin) [inline libs/genai/langchain_google_genai/_function_utils.py:66]: This should probably be before line # 57 :)

File: libs/genai/langchain_google_genai/_function_utils.py [hunk 4]
```diff
@@ -47,29 +42,36 @@ class _ToolDict(TypedDict):
...
+    tool: GeminiToolLike,
 ) -> GoogleTool:
     if isinstance(tool, GoogleTool):
         return cast(GoogleTool, tool)
-    if isinstance(tool, type) and issubclass(tool, BaseModel):
+    if isinstance(tool, (BaseTool, Type[BaseModel], dict, FunctionDeclaration)):
         return GoogleTool(function_declarations=[_convert_to_genai_function(tool)])
     if callable(tool):
-        return _convert_tool_to_genai_function(callable_as_lc_tool()(tool))
-    if isinstance(tool, list):
```
Comment (lkuligin) [inline libs/genai/langchain_google_genai/_function_utils.py:64]: we allow a list (a sequence), should we keep it?

Comment (nobu007) [reply libs/genai/langchain_google_genai/_function_utils.py:64]: Thanks. I fixed it.

File: libs/genai/langchain_google_genai/_function_utils.py (hunk 2 above)
Comment (nobu007) [reply libs/genai/langchain_google_genai/_function_utils.py]: I added some comments and dict key checking.

File: libs/genai/langchain_google_genai/_function_utils.py (hunk 1 above)
Comment (nobu007) [reply libs/genai/langchain_google_genai/_function_utils.py]: fixed it.

File: libs/genai/langchain_google_genai/_function_utils.py (hunk 3 above)
Comment (nobu007) [reply libs/genai/langchain_google_genai/_function_utils.py]: @lkuligin 
I think changed the order callable -> list -> dict.
Anyway, I improved code(also pytest passed). Please check again?

**Comment by lkuligin** [comment]
@nobu007 are you planning to finalize this one? there are a few linting errors, and the integration test is breaking.

**Comment by nobu007** [comment]
@lkuligin 
I fixed as I can.  But  linting errors still remains.
It may difficult all errrors fix in this PR.
Do you have any request for merge?

**Comment by lkuligin** [comment]
> @lkuligin I fixed as I can. But linting errors still remains. It may difficult all errrors fix in this PR. Do you have any request for merge?

we need to fix all remaining linting errors, linter is green on other ones

**Review by lkuligin — APPROVED** [review]

**Pull request merged**